feat: Add command line interface CLI

pke opened the pull request:
Provides a simple to use command line interface `json-mask` that filters a piped stream or input file.

Output can be redirected to a file or another command.

There is a problem with the commented out test. It will timeout. The culprit seems to be the shebang in `cli.js`. When removed the test runs and does not timeout (but fails of course). Any ideas? I'll create it as draft until I figured this out.

nemtsov commented:
@pke I love this! Thanks so much for putting it together. 

Your two main issues are: (a) you need to give the `cli.js` executable permissions (via `chmod 755 ./lib/cli.js`) and (b) `.finally()` isn't a thing in node v8 (it was introduced in v10). 

Apart from that, I see only a few minor improvements that can be made:
  1. Use `async / await`, since that's available. This should simplify things a bit.
  1. Use `require('util').promisify` on `fs.readFile`
  1. Use using arrow function expressions
  1. Consider keeping the cli scripts in a high-level `bin` folder with the file-name matching the executable (e.g., `mv ./lib/cli.js ./bin/json-mask`). I prefer to do  this to signal that those file shouldn't be `required()` and so they are easier to find.
  1. I don't believe you need to `JSON.stringify` before `console.log`ing

pke commented:
So, I have rewritten large portions according to your notes.

1. I have moved the cli as json-mask to the bin folder. Hopefully this will make npm chmod the file upon install
2. I removed `finally` it was not required anyway.

>Use async / await, since that's available. This should simplify things a bit.

I tried it, it made the code actually more cluttered and more LOC.
I have tightend up the promises a lot and the chaining now looks very clean.

> Use require('util').promisify on fs.readFile

:heavy_check_mark: I have used `const fs = require('fs').promises`

> Use using arrow function expressions

:heavy_check_mark: (where it made sense)
Usually when contributing to a repo I check which function style is used.
And I didn't see any arrow functions used in the existing code.

> Consider keeping the cli scripts in a high-level bin folder with the file-name matching the executable (e.g., mv ./lib/cli.js ./bin/json-mask). I prefer to do this to signal that those file shouldn't be required() and so they are easier to find.

Did that, and I hope thats also a hint to npm to `chmod +x` the file upon install. I mean it creates a .cmd file when installed on Windows, lets hope it works. I've checked other CLI packages and no one is doing a `postinstall: chmod x` anywhere. But all have their scripts in the `bin` folder.

> I don't believe you need to JSON.stringify before console.loging

Yes, you need. `console.log` does not print JSON when handing in an object. Try it ;)

I also introduced [nve](https://github.com/ehmicky/nve/blob/master/README.md) for testing on all 8, 10, 12 and latest node version without having to install them. Its a great tool and it immediately catched the `finally`when running the test against node 8. This also works in CI environments. If you want I can rewrite the `npm test` script to this:

```json
"scripts": {
  "test": "lint && nve 8 10 12 \"*\" mocha"
}
```

For now its runnable as `npm run test:all`

Once you approved this PR I'll rebase to the current master.

nemtsov commented on bin/json-mask.js:
```diff
@@ -0,0 +1,74 @@
+
+const maskInput = fields => {
+  if (!fields) {
+    throw new Error('Fields argument missing')
```
This error will be thrown without usage.

pke replied:
What do you mean? If no fields are given, the run command fails.

nemtsov replied:
Sorry I wasn’t clear. Unlike other errors this one will not print out the user-friendly usage info message and will show a stack trace.

nemtsov commented on bin/json-mask.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+function readInputFile (inputFile, resolve, reject) {
+  const fs = require('fs').promises
```
1. `require('fs').promises` is only available since Node v10. So this throws an error a few lines  below.
2. Let's put all `require()`s on top of the file. This will make it easier to transition to `import`s later.

Consider replacing this function altogether with the following on top of the file:

```js
const fs = require('fs')
const { promisify } = require('util')

const readFile = promisify(fs.readFile)
```

Btw, I'm drawing a blank on why `path.resolve` is necessary here. Can you remind me, please?

pke replied:
I have run the tests on using `nve 8` and it ran just fine. What gives?!
I'll change it though.

pke replied:
`path.resolve` resolves relative input paths to absolute ones. I am not really familiar with the node file functions, but isn't it better to hand in an absolute path to readFile just to be sure where the file is read from?

nemtsov replied:
I’ve tried both relative and absolute paths and readFile seems to work fine without a resolve, so I can’t seem to find a case where this is needed.

nemtsov commented on test/cli-test.js:
```diff
@@ -0,0 +1,70 @@
+  }
+  }, */
+  {
+    c: `cat ${__dirname}/fixture/activities.json | node ${__dirname}/../bin/json-mask.js "kind"`,
```
I didn't catch this before, but what will we do with our Windows friends? I'm concerned that they won't be able to run this test. Are there any alternatives?

pke replied:
I'll rewrite the tests with `echo` piping, should work on Windows too.
I'll also add Windows build env to travis.

nemtsov commented on bin/json-mask.js:
```diff
@@ -0,0 +1,74 @@
+  })
+}
+
+function readInputFile (inputFile, resolve, reject) {
```
`resolve` and `reject` are not used any more

nemtsov commented on bin/json-mask.js:
```diff
@@ -0,0 +1,74 @@
+
+run(process.argv[2], process.argv[3])
+  .then(writeOutput)
+  .catch(usage)
```
Consider, simplifying like this, for example:

```js
async function run (fields, inputFilePath) {
  const input = JSON.parse(await getInput(inputFilePath))
  const masked = mask(input, fields)
  console.log(JSON.stringify(masked))
}

run(process.argv[2], process.argv[3])
  .catch(usage)
```

Putting it inside `async / await` like in the example below, will also catch rogue Errors like on line 53 above. It also makes `parseInput`, `maskInput` and `writeOutput` unnecessary.

You were right about `console.log` and json, btw—`JSON.stringify` is still necessary; thanks for the reminder.

nemtsov commented on test/cli-test.js:
```diff
@@ -0,0 +1,70 @@
+      })
+    }(tests[i]))
+  }
+})
```
Overall the coverage is still pretty low on `bin/json-mask`, especially on branches. If you can, try covering as much as possible.

```
---------------|----------|----------|----------|----------|-------------------|
File           |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
---------------|----------|----------|----------|----------|-------------------|
All files      |    90.07 |    79.52 |    95.83 |    92.42 |                   |
 bin           |    64.29 |     12.5 |    85.71 |    61.54 |                   |
  json-mask.js |    64.29 |     12.5 |    85.71 |    61.54 |... 14,15,31,32,34 |
---------------|----------|----------|----------|----------|-------------------|
```

pke replied:
A sure, didn't have this on my radar. node-tap runs coverage for me automatically on each test. Will bring this up to 100%.

nemtsov commented on package.json:
```diff
@@ -5,6 +5,7 @@
   "main": "lib/index",
   "scripts": {
     "test": "npm run lint && mocha",
+    "test:all": "nve 8 10 12 \"*\" mocha",
```
I'd rather not do this just yet. My main concern is the current assumption is that the user has `nve` installed globally; which I'm not comfortable with. And I'm not sure whether it'd mean to bring this in as a dev dependency. Let's discuss this in a separate PR if you're passionate about it.

pke replied:
I think you might be confusing `nve` with `nvm`. nve is installed locally and can be used to run tests against different node version *without* installing them globally.
[Read](https://github.com/ehmicky/nve/blob/master/README.md) about it

ehmicky replied:
I can confirm `nve` is installed locally, not globally (I created that package).
If you need any support or have any questions, please let me know directly or on the GitHub repo :)

pke commented:
Hilarious, the JSON.parse spits out different error messages on Windows!
Gonna fix my error message checks then, and see why the `echo` test does not run on Windows.
Expect some more commits here, when merging you might considering squash ;)

pke commented:
What do you think @nemtsov? Looks good now or other change requests?

nemtsov commented:
@pke your feature is published as `json-mask@1.0.1`. Thank you for your patience.

pke commented:
Hurray for the nice feature addition. Thanks for merging and releasing a new version.
